packages/smithy-aws-core/src/smithy_aws_core/credentials_resolvers/imds.py

```diff
@@ -0,0 +1,230 @@
+        port: int = 80,
+        token_ttl: int = 21600,
+        ec2_instance_profile_name: str | None = None,
+    ):
+        self.retry_strategy = retry_strategy or SimpleRetryStrategy(max_attempts=3)
+        self.endpoint_mode = endpoint_mode
+        self.endpoint_uri = self._resolve_endpoint(endpoint_uri, endpoint_mode)
+        self.port = port
+        self.token_ttl = token_ttl
+        self.ec2_instance_profile_name = ec2_instance_profile_name
+
+    def _resolve_endpoint(
+        self, endpoint_uri: URI | None, endpoint_mode: Literal["IPv4", "IPv6"]
+    ) -> URI:
+        if endpoint_uri is not None:
+            return endpoint_uri
+
+        host_mapping = {"IPv4": "169.254.169.254", "IPv6": "[fd00:ec2::254]"}
+
+        return URI(
+            scheme="http", host=host_mapping.get(endpoint_mode, host_mapping["IPv4"])
+        )
+
+
+class EC2Metadata:
+    def __init__(self, http_client: HTTPClient, config: Config | None = None):
+        self._http_client = http_client
+        self._config = config or Config()
+        self._token_cache = TokenCache(
+            http_client=self._http_client,
+            base_uri=self._config.endpoint_uri,
+            token_ttl=self._config.token_ttl,
+        )
+
+    async def get(self, *, path: str) -> str:
+        token = await self._token_cache.get_token()
+        headers = Fields(
+            [
+                # TODO: Add user-agent
+                Field(
+                    name="x-aws-ec2-metadata-token",
+                    values=[token.value.decode("utf-8")],
+                )
+            ]
+        )
+        request = HTTPRequest(
+            method="GET",
+            destination=URI(
+                scheme=self._config.endpoint_uri.scheme,
+                host=self._config.endpoint_uri.host,
+                port=self._config.port,
+                path=path,
+            ),
+            fields=headers,
+        )
+        response = await self._http_client.send(request=request)
+        body = await response.consume_body_async()
+        return body.decode("utf-8")
+
+
+class IMDSCredentialsResolver(
+    IdentityResolver[AWSCredentialsIdentity, IdentityProperties]
+):
+    """Resolves AWS Credentials from an EC2 Instance Metadata Service (IMDS) client."""
+
+    # TODO: Handle fallback to legacy path when a 404 is received.
+    _METADATA_PATH_BASE = "/latest/meta-data/iam/security-credentials-extended/"
+
+    def __init__(self, http_client: HTTPClient, config: Config | None = None):
+        self._http_client = http_client
+        self._ec2_metadata_client = EC2Metadata(http_client=http_client, config=config)
+        self._config = config or Config()
+        self._credentials = None
+        self._profile_name = self._config.ec2_instance_profile_name
+
+    async def get_identity(
+        self, *, identity_properties: IdentityProperties
+    ) -> AWSCredentialsIdentity:
+        if self._credentials is not None:
+            return self._credentials
```

Comment:
The way IMDS works we should end up with the same credentials here if we get concurrent invocations, I'm just wondering if we need anything to control setting this. If this happens right on a push window is it possible we accidentally get older credentials?